### ISSUE TITLE
Improve datatype parsing error message

### ISSUE THREAD

**Issue by liss-h** [TortugaAttack/rdf4cpp#204]
Currently errors that occur during datatype parsing are often confusing or just not very good. This needs to be improved.

Addions by @bigerl:
The `ConstexprString datatype` parameter in some parsing functions does not enforce in any way to be set to correct values only. It's very easy to set it to some random value. I have no simple idea at hand but a more robust approach would be preferable.

In general a more robust way to generate messages would be preferable.

**Comment by KonradHoeffner** [comment]
The errors are also too many, for example when I pipe them into a file it has 21 MB of "invalid date..." messages after loading an 1.2GB Turtle file.

**Comment by liss-h** [comment]
> The errors are also too many, for example when I pipe them into a file it has 21 MB of "invalid date..." messages after loading an 1.2GB Turtle file.

Have you verified that the triples/dates in that file are actually correct? If so it would be very helpful if you could post them so we can see which triples we erroneously reject and fix it.

**Comment by KonradHoeffner** [comment]
The file in question is <http://downloads.dbpedia.org/2016-10/core-i18n/en/persondata_en.ttl.bz2> and the dates do seem partially wrong (some are missing the leading zero for months and days, others even have 0 for the month or day) but in this case there are 182774 errors, I think it would be enough to report the first 10 or so, to not slow the program down, or to allow users to ignore the errors somehow.

Some of the triples

```
<http://dbpedia.org/resource/Buddie_Petit> <http://dbpedia.org/ontology/birthDate> "1890-0-0"^^<http://www.w3.org/2001/XMLSchema#date> .
<http://dbpedia.org/resource/Buddy_Emmons> <http://dbpedia.org/ontology/deathDate> "2015-7-0"^^<http://www.w3.org/2001/XMLSchema#date> .
<http://dbpedia.org/resource/Margaret_of_Foix-Candale> <http://dbpedia.org/ontology/birthDate> "1473-0-0"^^<http://www.w3.org/2001/XMLSchema#date> .
```

Having month or day as 0 seems indeed wrong, I guess the authors wanted to express that they didn't know the exact date.
So in summary, the triples are correct RDF but the datatypes are invalid dates, which are rightfully reported as incorrect, but from the data consumption side (e.g. using DBpedia where this is from) I am using the data as is so it would be nice if the errors could be silenced or abbreviated.

**Comment by liss-h** [comment]
I mean, rdf4cpp is a library so you can make your program do whatever you want. 

Example, only print the first 10 errors:
```c++
using namespace rdf4cpp::rdf;
parser::ParsingFlags flags = parser::ParsingFlag::Turtle;

size_t errors = 0;

for (const auto &e: parser::RDFFileParser{filename, flags}) {
  if (e.has_value()) {
    // do stuff with triple
  } else if (errors++ < 10) {
      std::cerr << e.error() << std::endl;
  }
}
```
You can even manually inspect the [error object](https://github.com/rdf4cpp/rdf4cpp/blob/a3588b19b6fc5cb11717a32eb0a69491acf70b27/src/rdf4cpp/rdf/parser/ParsingError.hpp#L13) an serialize it to the terminal however you want.

**Comment by KonradHoeffner** [comment]
Oh sorry, I didn't realize the error messages were initiated from within our benchmark code and not from within the rdf4cpp library itself, you are right, then I withdraw my suggestion, thanks for the advice!

**Comment by liss-h** [comment]
> Oh sorry, I didn't realize the error messages were initiated from within our benchmark code and not from within the rdf4cpp library itself, you are right, then I withdraw my suggestion, thanks for the advice!

No worries 🙂
